README.md
# Barrels of Steam
###### A Factorio mod - [get it here!](https://mods.factorio.com/mod/barrels-of-steam)

![Mod Thumbnail](src/thumbnail.png "Mod Thumbnail")

This mod adds steam barrels at 165 °C and 500 °C, as well as recipes for barreling and unbarreling.

This all stems from a weird plan a friend and I had in our game to build a nuclear reactor on a remote island that we never set foot on, using only drones to send fuel rods across the sea and carry back superheated steam in barrels. Yes, it's a little silly, but you can put any fluid in a tank but not every fluid in a *barrel*, and we had just kind of expected this feature to already be there, and it wasn't probably due to temperature being relevant.

Anyway, our next game plan is to create a base powered only by barrels of hot steam - no solar, no big pylons, no pipes, no tanks, plus more.

## Recipes

### Steam Barrel

**Empty barrel** + 50× **Steam** (≥ 165 °C) ⇌ **Steam barrel** (165 °C)

Unlocked by **Fluid handling** technology.

![Fill Steam Barrel recipe](images/fill-steam-barrel-recipe.png "Fill Steam Barrel recipe")
![Empty Steam Barrel recipe](images/empty-steam-barrel-recipe.png "Empty Steam Barrel recipe")
![Fluid Handling technology](images/fluid-handling-technology.png "Fluid Handling technology")

### Superheated Steam Barrel

**Empty barrel** + 50× **Steam** (≥ 500 °C) ⇌ **Superheated steam barrel** (500 °C)

Unlocked by **Nuclear power** technology.

![Fill Superheated Steam Barrel recipe](images/fill-superheated-steam-barrel-recipe.png "Fill Superheated Steam Barrel recipe")
![Empty Superheated Steam Barrel recipe](images/empty-superheated-steam-barrel-recipe.png "Empty Superheated Steam Barrel recipe")
![Nuclear Power technology](images/nuclear-power-technology.png "Nuclear Power technology")

## Installation

It's simplest to just use the in-game mod loader to search for "Barrels of Steam" to get the latest published release.

But you can install a development version by copying `barrels-of-steam_x.x.x.zip` to [Factorio's mods folder](https://wiki.factorio.com/index.php?title=Application_directory):
- Windows `%appdata%\Factorio\mods`
- Mac `~/Library/Application Support/factorio/mods`
- Linux `~/.factorio/mods`

Running the build script `build.ps1` in PowerShell zips up `src` and does this automatically (Windows only, 7-zip is required on your path).

Otherwise, copy a zip from `build/<factorio_version>/barrels-of-steam_x.x.x.zip`, or download from [mods.factorio.com](https://mods.factorio.com/mod/barrels-of-steam), and paste it to Factorio's mods folder:


src/changelog.txt
---------------------------------------------------------------------------------------------------
Version: 18.0.6
Date: 12 Dec 2020
  Changes:
    - Added support for Factorio v1.1.
    - Adjusted fill recipe icons to match the new 1.1 ones.
---------------------------------------------------------------------------------------------------
Version: 17.0.5 / 18.0.5
Date: 20 Apr 2020
  Changes:
    - Renamed all recipes and items to support certain combinations of other mods.
---------------------------------------------------------------------------------------------------
Version: 17.0.4 / 18.0.4
Date: 19 Apr 2020
  Changes:
    - Added optional 15 °C ambient temperature steam barrels to support certain other mods.
    - Ambient steam recipes can be enabled in the in-game mod settings (disabled by default).
---------------------------------------------------------------------------------------------------
Version: 17.0.3 / 18.0.3
Date: 05 Apr 2020
  Bugfixes:
    - Recipes now unlock if the mod was installed after the vanilla technologies were researched.
---------------------------------------------------------------------------------------------------
Version: 17.0.2 / 18.0.2
Date: 02 Apr 2020
  Changes:
    - Added support for Factorio v0.18 by detecting correct icon sizes between versions.
---------------------------------------------------------------------------------------------------
Version: 0.0.1
Date: 29 Mar 2020
  Changes:
    - Initial release for Factorio v0.17.
    - Added 165 °C and 500 °C steam barrels, plus recipes and technologies.
---------------------------------------------------------------------------------------------------


build.ps1
# Build Script
# This script requires 7-zip (7z.exe) to be on your path

# Need name, version, and factorio_version from info.json
$info = Get-Content '.\src\info.json' | Out-String | ConvertFrom-Json

# Mod must go in a folder named with this specific format
$folder = "$($info.name)_$($info.version)"
echo $folder

# Need an intermediate folder to arrange files before zipping to the output
$intermediate = ".\build\$folder"
$output = ".\build\$($info.factorio_version)\$folder.zip"

# Copy src to intermediate to be zipped to output, then clear up after
Copy-Item '.\src' -Destination $intermediate -Recurse
#Compress-Archive -Path $intermediate -Force -DestinationPath $output
7z a $output $intermediate\
Remove-Item -Path $intermediate -Recurse

# Install output zip to mods folder
Copy-Item $output -Destination "$($env:APPDATA)\Factorio\mods"


src/control.lua
-- Unlock recipes if mod was installed after technology was researched
script.on_configuration_changed(function(e)
  for _, force in pairs(game.forces) do
    -- REGULAR STEAM BARREL
    local hasFluidHandling = force.technologies["fluid-handling"].researched
    force.recipes["fill-steam-barrel-165"].enabled  = hasFluidHandling
    force.recipes["empty-steam-barrel-165"].enabled = hasFluidHandling
    
    -- SUPERHEATED NUCLEAR STEAM BARREL
    local hasNuclearPower = force.technologies["nuclear-power"].researched
    force.recipes["fill-steam-barrel-500"].enabled  = hasNuclearPower
    force.recipes["empty-steam-barrel-500"].enabled = hasNuclearPower

    -- UNINSULATED AMBIENT STEAM BARREL
    local hasAmbientBarrels = settings.startup["enable-ambient-steam-barrel"].value and hasFluidHandling
    force.recipes["fill-steam-barrel-015"].enabled  = hasAmbientBarrels
    force.recipes["empty-steam-barrel-015"].enabled = hasAmbientBarrels
  end
end)


src/prototypes/steam-barrel.lua
------------------
-- STEAM BARREL --
------------------

local iconSize = data.raw.item["empty-barrel"].icon_size -- 32px in v0.17, 64px in v0.18

local steamBarrel165 = {
  icon_mipmaps = 4,
  icon_size = 64,
  icons = {
    {
      icon = "__base__/graphics/icons/fluid/barreling/empty-barrel.png",
      icon_mipmaps = 4,
      icon_size = iconSize -- 32px in v0.17, 64px in v0.18
    },
    {
      icon = "__base__/graphics/icons/fluid/barreling/barrel-side-mask.png",
      icon_mipmaps = 4,
      icon_size = iconSize, -- 32px in v0.17, 64px in v0.18
      tint = {
        a = 0.75,
        b = 0.6,
        g = 0.34,
        r = 0.7 -- was 0.0 for water barrels
      }
    },
    {
      icon = "__base__/graphics/icons/fluid/barreling/barrel-hoop-top-mask.png",
      icon_mipmaps = 4,
      icon_size = iconSize, -- 32px in v0.17, 64px in v0.18
      tint = {
        a = 0.75,
        b = 0.7,
        g = 0.7,
        r = 0.7 -- was 0.0 for water barrels
      }
    }
  },
  localised_name = {
    "item-name.filled-barrel",
    {
      "fluid-name.steam"
    }
  },
  name = "steam-barrel-165",
  order = "b[steam-barrel-165]",
  stack_size = 10,
  subgroup = "fill-barrel",
  type = "item"
}

-- FILL STEAM BARREL RECIPE --

local fillSteamRecipe165 = {
  allow_decomposition = false,
  category = "crafting-with-fluid",
  enabled = false, -- can set to true to ignore tech unlocks
  energy_required = 0.2,
  icon_mipmaps = 4,
  icon_size = 64,
  icons = {
    {
      icon = "__base__/graphics/icons/fluid/barreling/barrel-fill.png",
      icon_mipmaps = 4,
      icon_size = iconSize -- 32px in v0.17, 64px in v0.18
    },
    {
      icon = "__base__/graphics/icons/fluid/barreling/barrel-fill-side-mask.png",
      icon_mipmaps = 4,
      icon_size = iconSize, -- 32px in v0.17, 64px in v0.18
      tint = {
        a = 0.75,
        b = 0.6,
        g = 0.34,
        r = 0.7 -- was 0.0 for water barrels
      }
    },
    {
      icon = "__base__/graphics/icons/fluid/barreling/barrel-fill-top-mask.png",
      icon_mipmaps = 4,
      icon_size = iconSize, -- 32px in v0.17, 64px in v0.18
      tint = {
        a = 0.75,
        b = 0.7,
        g = 0.7,
        r = 0.7
      }
    },
    {
      icon = "__base__/graphics/icons/fluid/steam.png",
      icon_mipmaps = 4,
      icon_size = iconSize,  -- 32px in v0.17, 64px in v0.18
      scale = 16 / iconSize, -- 0.5  in v0.17, 0.25 in v0.18
      shift = {
        -8,
        -8
      }
    }
  },
  ingredients = {
    {
      amount = 50,
      catalyst_amount = 50,
      name = "steam",
      type = "fluid",
      minimum_temperature = 165.0
    },
    {
      amount = 1,
      catalyst_amount = 1,
      name = "empty-barrel",
      type = "item"
    }
  },
  localised_name = {
    "recipe-name.fill-barrel",
    {
      "fluid-name.steam"
    }
  },
  name = "fill-steam-barrel-165",
  order = "b[fill-steam-barrel-165]",
  results = {
    {
      amount = 1,
      catalyst_amount = 1,
      name = "steam-barrel-165",
      type = "item"
    }
  },
  subgroup = "fill-barrel",
  type = "recipe"
}

-- EMPTY STEAM BARREL RECIPE --

local emptySteamRecipe165 = {
  allow_decomposition = false,
  category = "crafting-with-fluid",
  enabled = false, -- can set to true to ignore tech unlocks
  energy_required = 0.2,
  icon_mipmaps = 4,
  icon_size = 64,
  icons = {
    {
      icon = "__base__/graphics/icons/fluid/barreling/barrel-empty.png",
      icon_mipmaps = 4,
      icon_size = iconSize -- 32px in v0.17, 64px in v0.18
    },
    {
      icon = "__base__/graphics/icons/fluid/barreling/barrel-empty-side-mask.png",
      icon_mipmaps = 4,
      icon_size = iconSize, -- 32px in v0.17, 64px in v0.18
      tint = {
        a = 0.75,
        b = 0.6,
        g = 0.34,
        r = 0.7 -- was 0.0 for water barrels
      }
    },
    {
      icon = "__base__/graphics/icons/fluid/barreling/barrel-empty-top-mask.png",
      icon_mipmaps = 4,
      icon_size = iconSize, -- 32px in v0.17, 64px in v0.18
      tint = {
        a = 0.75,
        b = 0.7,
        g = 0.7,
        r = 0.7
      }
    },
    {
      icon = "__base__/graphics/icons/fluid/steam.png",
      icon_mipmaps = 4,
      icon_size = iconSize,  -- 32px in v0.17, 64px in v0.18
      scale = 16 / iconSize, -- 0.5  in v0.17, 0.25 in v0.18
      shift = {
        7,
        8
      }
    }
  },
  ingredients = {
    {
      amount = 1,
      catalyst_amount = 1,
      name = "steam-barrel-165",
      type = "item"
    }
  },
  localised_name = {
    "recipe-name.empty-filled-barrel",
    {
      "fluid-name.steam"
    }
  },
  name = "empty-steam-barrel-165",
  order = "c[empty-steam-barrel-165]",
  results = {
    {
      amount = 50,
      catalyst_amount = 50,
      name = "steam",
      type = "fluid",
      temperature = 165
    },
    {
      amount = 1,
      catalyst_amount = 1,
      name = "empty-barrel",
      type = "item"
    }
  },
  subgroup = "empty-barrel",
  type = "recipe"
}

-- TECHNOLOGY --

table.insert(data.raw.technology["fluid-handling"].effects, {type = "unlock-recipe", recipe = "fill-steam-barrel-165"})
table.insert(data.raw.technology["fluid-handling"].effects, {type = "unlock-recipe", recipe = "empty-steam-barrel-165"})

-- EXTEND --

data:extend{
  steamBarrel165,
  fillSteamRecipe165,
  emptySteamRecipe165
}

---------------------------------
-- SUPERHEATED NUCLEAR VARIANT --
---------------------------------

local steamBarrel500 = table.deepcopy(steamBarrel165)
steamBarrel500.name = "steam-barrel-500"
steamBarrel500.order = "b[steam-barrel-500]"
steamBarrel500.localised_name[2][1] = "fluid-name.super-steam"
steamBarrel500.icons[2].tint.r = 1.0
steamBarrel500.icons[3].tint.r = 1.0

local fillSteamRecipe500 = table.deepcopy(fillSteamRecipe165)
fillSteamRecipe500.name = "fill-steam-barrel-500"
fillSteamRecipe500.order = "b[fill-steam-barrel-500]"
fillSteamRecipe500.localised_name[2][1] = "fluid-name.super-steam"
fillSteamRecipe500.icons[2].tint.r = 1.0
fillSteamRecipe500.ingredients[1].minimum_temperature = 500.0
fillSteamRecipe500.results[1].name = "steam-barrel-500"

local emptySteamRecipe500 = table.deepcopy(emptySteamRecipe165)
emptySteamRecipe500.name = "empty-steam-barrel-500"
emptySteamRecipe500.order = "c[empty-steam-barrel-500]"
emptySteamRecipe500.localised_name[2][1] = "fluid-name.super-steam"
emptySteamRecipe500.icons[2].tint.r = 1.0
emptySteamRecipe500.ingredients[1].name = "steam-barrel-500"
emptySteamRecipe500.results[1].temperature = 500.0

-- TECHNOLOGY --

table.insert(data.raw.technology["nuclear-power"].effects, {type = "unlock-recipe", recipe = "fill-steam-barrel-500"})
table.insert(data.raw.technology["nuclear-power"].effects, {type = "unlock-recipe", recipe = "empty-steam-barrel-500"})

-- EXTEND --

data:extend{
  steamBarrel500,
  fillSteamRecipe500,
  emptySteamRecipe500
}

---------------------------------
-- UNINSULATED AMBIENT VARIANT --
---------------------------------

local steamBarrel015 = table.deepcopy(steamBarrel165)
steamBarrel015.name = "steam-barrel-015"
steamBarrel015.order = "b[steam-barrel-015]"
steamBarrel015.localised_name[2][1] = "fluid-name.ambient-steam"
steamBarrel015.icons[2].tint.r = 0.3

local fillSteamRecipe015 = table.deepcopy(fillSteamRecipe165)
fillSteamRecipe015.name = "fill-steam-barrel-015"
fillSteamRecipe015.order = "b[fill-steam-barrel-015]"
fillSteamRecipe015.localised_name[2][1] = "fluid-name.ambient-steam"
fillSteamRecipe015.icons[2].tint.r = 0.3
fillSteamRecipe015.ingredients[1].minimum_temperature = nil
fillSteamRecipe015.results[1].name = "steam-barrel-015"

local emptySteamRecipe015 = table.deepcopy(emptySteamRecipe165)
emptySteamRecipe015.name = "empty-steam-barrel-015"
emptySteamRecipe015.order = "c[empty-steam-barrel-015]"
emptySteamRecipe015.localised_name[2][1] = "fluid-name.ambient-steam"
emptySteamRecipe015.icons[2].tint.r = 0.3
emptySteamRecipe015.ingredients[1].name = "steam-barrel-015"
emptySteamRecipe015.results[1].temperature = 15.0

-- TECHNOLOGY --

if (settings.startup["enable-ambient-steam-barrel"].value) then
  table.insert(data.raw.technology["fluid-handling"].effects, {type = "unlock-recipe", recipe = "fill-steam-barrel-015"})
  table.insert(data.raw.technology["fluid-handling"].effects, {type = "unlock-recipe", recipe = "empty-steam-barrel-015"})
end

-- EXTEND --

data:extend{
  steamBarrel015,
  fillSteamRecipe015,
  emptySteamRecipe015
}
